llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp

```diff
@@ -1336,33 +1355,86 @@ void AsmPrinter::emitFrameAlloc(const MachineInstr &MI) {
+    OutStreamer.AddComment("function entry count");
+    OutStreamer.emitULEB128IntValue(
+        llvm::transformOptional(MF.getFunction().getEntryCount(),
+                                std::mem_fn(&Function::ProfileCount::getCount))
+            .value_or(0));
+  }
+}
+
+static void emitPGOBBData(const MachineFunction &MF,
+                          const MachineBasicBlock &MBB,
+                          const MachineBlockFrequencyInfo &MBFI,
+                          const MachineBranchProbabilityInfo &MBPI,
+                          MCStreamer &OutStreamer) {
+  bool UseBrProb = PgoBBAddrMapFeatures.isSet(PGOMapFeaturesEnum::BrProb);
+  unsigned SuccCount;
+  auto SuccsType = object::PGOBBAddrMap::BBEntry::SuccessorsType::None;
+  if (UseBrProb) {
+    SuccCount = MBB.succ_size();
+    SuccsType = object::PGOBBAddrMap::BBEntry::getSuccessorsType(SuccCount);
+  }
+
+  // Must place metadata before extra info
+  OutStreamer.emitULEB128IntValue(object::PGOBBAddrMap::BBEntry::encodeMD(
+      getBBAddrMapMetadata(MBB), SuccsType));
+
+  if (PgoBBAddrMapFeatures.isSet(PGOMapFeaturesEnum::BBFreq)) {
+    OutStreamer.AddComment("basic block frequency");
+    OutStreamer.emitULEB128IntValue(MBFI.getBlockFreq(&MBB).getFrequency());
+  }
+
+  if (UseBrProb) {
+    if (SuccsType == object::PGOBBAddrMap::BBEntry::SuccessorsType::Multiple) {
+      OutStreamer.AddComment("basic block successor count");
+      OutStreamer.emitULEB128IntValue(SuccCount);
+    }
+    for (const MachineBasicBlock *SuccMBB : MBB.successors()) {
+      OutStreamer.AddComment("successor BB ID");
+      OutStreamer.emitULEB128IntValue(SuccMBB->getBBID()->BaseID);
+      OutStreamer.AddComment("successor branch probability");
+      OutStreamer.emitULEB128IntValue(
+          MBPI.getEdgeProbability(&MBB, SuccMBB).getNumerator());
+    }
+  }
 }
 
 void AsmPrinter::emitBBAddrMapSection(const MachineFunction &MF) {
-  MCSection *BBAddrMapSection =
-      getObjFileLowering().getBBAddrMapSection(*MF.getSection());
+  unsigned PGOFeatures = PgoBBAddrMapFeatures.getBits();
+  bool UsePGOExtension = PGOFeatures != 0;
+
+  MCSection *BBAddrMapSection = getObjFileLowering().getBBAddrMapSection(
+      *MF.getSection(), UsePGOExtension);
   assert(BBAddrMapSection && ".llvm_bb_addr_map section is not initialized.");
 
   const MCSymbol *FunctionSymbol = getFunctionBegin();
+  uint8_t BBAddrMapVersion = OutStreamer->getContext().getBBAddrMapVersion();
+  assert((UsePGOExtension ? BBAddrMapVersion >= 2 : true) &&
+         "PGOBBAddrMap only supports version 2 or later");
 
   OutStreamer->pushSection();
   OutStreamer->switchSection(BBAddrMapSection);
   OutStreamer->AddComment("version");
-  uint8_t BBAddrMapVersion = OutStreamer->getContext().getBBAddrMapVersion();
   OutStreamer->emitInt8(BBAddrMapVersion);
   OutStreamer->AddComment("feature");
-  OutStreamer->emitInt8(0);
+  OutStreamer->emitInt8(uint64_t(PGOFeatures));
```

Comment:
why cast to `uint64_t` when calling `emitInt8`? Wouldn't this make more sense as `static_cast<uint8_t>(PGOFeatures)`?

Reply:
Probably just auto-piloted after seeing that emitInt8 takes in uint64_t. Changed.